File: HEUR-Trojan/Win32/G/HEUR-Trojan.Win32.Generic-8862636861e0bb63dbb1c90a72bac6a40746efa23d007ced52614a655741a650/svchst/Costura/AssemblyLoader.vb
Imports System
Imports System.Collections.Generic
Imports System.Globalization
Imports System.IO
Imports System.IO.Compression
Imports System.Reflection
Imports System.Runtime.CompilerServices
Imports System.Threading

Namespace Costura
	' Token: 0x02000076 RID: 118
	<CompilerGenerated()>
	Friend Module AssemblyLoader
		' Token: 0x06000211 RID: 529 RVA: 0x00002B6F File Offset: 0x00000D6F
		Private Function CultureToString(culture As CultureInfo) As String
			If culture Is Nothing Then
				Return ""
			End If
			Return culture.Name
		End Function

		' Token: 0x06000212 RID: 530 RVA: 0x00012280 File Offset: 0x00010480
		Private Function ReadExistingAssembly(name As AssemblyName) As Assembly
			For Each assembly As Assembly In AppDomain.CurrentDomain.GetAssemblies()
				Dim name2 As AssemblyName = assembly.GetName()
				If String.Equals(name2.Name, name.Name, StringComparison.InvariantCultureIgnoreCase) AndAlso String.Equals(AssemblyLoader.CultureToString(name2.CultureInfo), AssemblyLoader.CultureToString(name.CultureInfo), StringComparison.InvariantCultureIgnoreCase) Then
					Return assembly
				End If
			Next
			Return Nothing
		End Function

		' Token: 0x06000213 RID: 531 RVA: 0x000122E8 File Offset: 0x000104E8
		Private Sub CopyTo(source As Stream, destination As Stream)
			Dim array As Byte() = New Byte(81919) {}
			While True
				Dim num As Integer = source.Read(array, 0, array.Length)
				Dim num2 As Integer = num
				If num = 0 Then
					Exit For
				End If
				destination.Write(array, 0, num2)
			End While
		End Sub

		' Token: 0x06000214 RID: 532 RVA: 0x0001231C File Offset: 0x0001051C
		Private Function LoadStream(fullName As String) As Stream
			Dim executingAssembly As Assembly = Assembly.GetExecutingAssembly()
			If fullName.EndsWith(".compressed") Then
				Dim stream As Stream
				Using manifestResourceStream As Stream = executingAssembly.GetManifestResourceStream(fullName)
					Using deflateStream As DeflateStream = New DeflateStream(manifestResourceStream, CompressionMode.Decompress)
						Dim memoryStream As MemoryStream = New MemoryStream()
						AssemblyLoader.CopyTo(deflateStream, memoryStream)
						memoryStream.Position = 0L
						stream = memoryStream
					End Using
				End Using
				Return stream
			End If
			Return executingAssembly.GetManifestResourceStream(fullName)
		End Function

		' Token: 0x06000215 RID: 533 RVA: 0x000123A8 File Offset: 0x000105A8
		Private Function LoadStream(resourceNames As Dictionary(Of String, String), name As String) As Stream
			Dim text As String
			If resourceNames.TryGetValue(name, text) Then
				Return AssemblyLoader.LoadStream(text)
			End If
			Return Nothing
		End Function

		' Token: 0x06000216 RID: 534 RVA: 0x000123C8 File Offset: 0x000105C8
		Private Function ReadStream(stream As Stream) As Byte()
			Dim array As Byte() = New Byte(stream.Length - 1) {}
			stream.Read(array, 0, array.Length)
			Return array
		End Function

		' Token: 0x06000217 RID: 535 RVA: 0x000123F0 File Offset: 0x000105F0
		Private Function ReadFromEmbeddedResources(assemblyNames As Dictionary(Of String, String), symbolNames As Dictionary(Of String, String), requestedAssemblyName As AssemblyName) As Assembly
			Dim text As String = requestedAssemblyName.Name.ToLowerInvariant()
			If requestedAssemblyName.CultureInfo IsNot Nothing AndAlso Not String.IsNullOrEmpty(requestedAssemblyName.CultureInfo.Name) Then
				text = requestedAssemblyName.CultureInfo.Name + "." + text
			End If
			Dim array As Byte()
			Using stream As Stream = AssemblyLoader.LoadStream(assemblyNames, text)
				If stream Is Nothing Then
					Return Nothing
				End If
				array = AssemblyLoader.ReadStream(stream)
			End Using
			Using stream2 As Stream = AssemblyLoader.LoadStream(symbolNames, text)
				If stream2 IsNot Nothing Then
					Dim array2 As Byte() = AssemblyLoader.ReadStream(stream2)
					Return Assembly.Load(array, array2)
				End If
			End Using
			Return Assembly.Load(array)
		End Function

		' Token: 0x06000218 RID: 536 RVA: 0x000124B0 File Offset: 0x000106B0
		Public Function ResolveAssembly(sender As Object, e As ResolveEventArgs) As Assembly
			Dim obj As Object = AssemblyLoader.nullCacheLock
			SyncLock obj
				If AssemblyLoader.nullCache.ContainsKey(e.Name) Then
					Return Nothing
				End If
			End SyncLock
			Dim assemblyName As AssemblyName = New AssemblyName(e.Name)
			Dim assembly As Assembly = AssemblyLoader.ReadExistingAssembly(assemblyName)
			If assembly IsNot Nothing Then
				Return assembly
			End If
			assembly = AssemblyLoader.ReadFromEmbeddedResources(AssemblyLoader.assemblyNames, AssemblyLoader.symbolNames, assemblyName)
			If assembly Is Nothing Then
				obj = AssemblyLoader.nullCacheLock
				SyncLock obj
					AssemblyLoader.nullCache(e.Name) = True
				End SyncLock
				If(assemblyName.Flags And AssemblyNameFlags.Retargetable) <> AssemblyNameFlags.None Then
					assembly = Assembly.Load(assemblyName)
				End If
			End If
			Return assembly
		End Function

		' Token: 0x06000219 RID: 537 RVA: 0x00012590 File Offset: 0x00010790
		' Note: this type is marked as 'beforefieldinit'.
		Shared Sub New()
			AssemblyLoader.assemblyNames.Add("ionic.zip", "costura.ionic.zip.dll.compressed")
			AssemblyLoader.assemblyNames.Add("newtonsoft.json", "costura.newtonsoft.json.dll.compressed")
			AssemblyLoader.assemblyNames.Add("skinsoft.visualstyler", "costura.skinsoft.visualstyler.dll.compressed")
			AssemblyLoader.assemblyNames.Add("system.management.automation", "costura.system.management.automation.dll.compressed")
		End Sub

		' Token: 0x0600021A RID: 538 RVA: 0x00002B80 File Offset: 0x00000D80
		Public Sub Attach()
			If Interlocked.Exchange(AssemblyLoader.isAttached, 1) = 1 Then
				Return
			End If
			AddHandler AppDomain.CurrentDomain.AssemblyResolve, AddressOf AssemblyLoader.ResolveAssembly
		End Sub

		' Token: 0x040001CA RID: 458
		Private nullCacheLock As Object = New Object()

		' Token: 0x040001CB RID: 459
		Private nullCache As Dictionary(Of String, Boolean) = New Dictionary(Of String, Boolean)()

		' Token: 0x040001CC RID: 460
		Private assemblyNames As Dictionary(Of String, String) = New Dictionary(Of String, String)()

		' Token: 0x040001CD RID: 461
		Private symbolNames As Dictionary(Of String, String) = New Dictionary(Of String, String)()

		' Token: 0x040001CE RID: 462
		Private isAttached As Integer
	End Module
End Namespace

File: HEUR-Trojan/Win32/G/HEUR-Trojan.Win32.Generic-e77392cb1ebe598a96caa2e6bc6c66cf0f9d8d56a9c53e9382d352a6e4f103ea/Costura/AssemblyLoader.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.IO.Compression;
using System.Reflection;
using System.Runtime.CompilerServices;

namespace Costura;

[CompilerGenerated]
internal static class AssemblyLoader
{
	private static readonly Dictionary<string, bool> nullCache = new Dictionary<string, bool>();

	private static readonly Dictionary<string, string> assemblyNames = new Dictionary<string, string>();

	private static readonly Dictionary<string, string> symbolNames = new Dictionary<string, string>();

	private static ResolveEventHandler CS_0024_003C_003E9__CachedAnonymousMethodDelegate1;

	private static string CultureToString(CultureInfo culture)
	{
		if (culture == null)
		{
			return "";
		}
		return AssemblyLoader.smethod_0(culture);
	}

	private static Assembly ReadExistingAssembly(AssemblyName name)
	{
		AppDomain appDomain_ = AssemblyLoader.smethod_1();
		Assembly[] array = AssemblyLoader.smethod_2(appDomain_);
		Assembly[] array2 = array;
		int num = 0;
		Assembly assembly;
		while (true)
		{
			if (num < array2.Length)
			{
				assembly = array2[num];
				AssemblyName assemblyName_ = AssemblyLoader.smethod_3(assembly);
				if (AssemblyLoader.smethod_5(AssemblyLoader.smethod_4(assemblyName_), AssemblyLoader.smethod_4(name), StringComparison.InvariantCultureIgnoreCase) && AssemblyLoader.smethod_5(CultureToString(AssemblyLoader.smethod_6(assemblyName_)), CultureToString(AssemblyLoader.smethod_6(name)), StringComparison.InvariantCultureIgnoreCase))
				{
					break;
				}
				num++;
				continue;
			}
			return null;
		}
		return assembly;
	}

	private static void CopyTo(Stream source, Stream destination)
	{
		byte[] array = new byte[81920];
		int int_;
		while ((int_ = AssemblyLoader.smethod_8(source, array, 0, array.Length)) != 0)
		{
			AssemblyLoader.smethod_7(destination, array, 0, int_);
		}
	}

	private static Stream LoadStream(string fullname)
	{
		Assembly assembly_ = AssemblyLoader.smethod_9();
		if (AssemblyLoader.smethod_10(fullname, ".zip"))
		{
			Stream stream = AssemblyLoader.smethod_11(assembly_, fullname);
			try
			{
				DeflateStream deflateStream = AssemblyLoader.smethod_12(stream, CompressionMode.Decompress);
				try
				{
					MemoryStream memoryStream = AssemblyLoader.smethod_13();
					CopyTo(deflateStream, memoryStream);
					AssemblyLoader.smethod_14((Stream)memoryStream, 0L);
					return memoryStream;
				}
				finally
				{
					if (deflateStream != null)
					{
						AssemblyLoader.smethod_15((IDisposable)deflateStream);
					}
				}
			}
			finally
			{
				if (stream != null)
				{
					AssemblyLoader.smethod_15((IDisposable)stream);
				}
			}
		}
		return AssemblyLoader.smethod_11(assembly_, fullname);
	}

	private static Stream LoadStream(Dictionary<string, string> resourceNames, string name)
	{
		if (resourceNames.TryGetValue(name, out var value))
		{
			return LoadStream(value);
		}
		return null;
	}

	private static byte[] ReadStream(Stream stream)
	{
		byte[] array = new byte[AssemblyLoader.smethod_16(stream)];
		AssemblyLoader.smethod_8(stream, array, 0, array.Length);
		return array;
	}

	private static Assembly ReadFromEmbeddedResources(Dictionary<string, string> assemblyNames, Dictionary<string, string> symbolNames, AssemblyName requestedAssemblyName)
	{
		string text = AssemblyLoader.smethod_17(AssemblyLoader.smethod_4(requestedAssemblyName));
		if (AssemblyLoader.smethod_6(requestedAssemblyName) != null && !AssemblyLoader.smethod_18(AssemblyLoader.smethod_0(AssemblyLoader.smethod_6(requestedAssemblyName))))
		{
			text = AssemblyLoader.smethod_19("{0}.{1}", (object)AssemblyLoader.smethod_0(AssemblyLoader.smethod_6(requestedAssemblyName)), (object)text);
		}
		Stream stream = LoadStream(assemblyNames, text);
		byte[] byte_;
		try
		{
			if (stream == null)
			{
				return null;
			}
			byte_ = ReadStream(stream);
		}
		finally
		{
			if (stream != null)
			{
				AssemblyLoader.smethod_15((IDisposable)stream);
			}
		}
		Stream stream2 = LoadStream(symbolNames, text);
		try
		{
			if (stream2 != null)
			{
				byte[] byte_2 = ReadStream(stream2);
				return AssemblyLoader.smethod_20(byte_, byte_2);
			}
		}
		finally
		{
			if (stream2 != null)
			{
				AssemblyLoader.smethod_15((IDisposable)stream2);
			}
		}
		return AssemblyLoader.smethod_21(byte_);
	}

	public static Assembly ResolveAssembly(string assemblyName)
	{
		if (nullCache.ContainsKey(assemblyName))
		{
			return null;
		}
		AssemblyName assemblyName2 = AssemblyLoader.smethod_22(assemblyName);
		Assembly assembly = ReadExistingAssembly(assemblyName2);
		if ((object)assembly != null)
		{
			return assembly;
		}
		assembly = ReadFromEmbeddedResources(assemblyNames, symbolNames, assemblyName2);
		if ((object)assembly == null)
		{
			nullCache.Add(assemblyName, value: true);
			if (AssemblyLoader.smethod_23(assemblyName2) == AssemblyNameFlags.Retargetable)
			{
				assembly = AssemblyLoader.smethod_24(assemblyName2);
			}
		}
		return assembly;
	}

	static AssemblyLoader()
	{
		assemblyNames.Add("newtonsoft.json", "costura.newtonsoft.json.dll.zip");
	}

	private static Assembly _003CAttach_003Eb__0(object s, ResolveEventArgs e)
	{
		return ResolveAssembly(AssemblyLoader.smethod_25(e));
	}

	public static void Attach()
	{
		AppDomain appDomain_ = AssemblyLoader.smethod_1();
		AssemblyLoader.smethod_26(appDomain_, (ResolveEventHandler)((object s, ResolveEventArgs e) => ResolveAssembly(AssemblyLoader.smethod_25(e))));
	}

	static string smethod_0(CultureInfo cultureInfo_0)
	{
		return cultureInfo_0.Name;
	}

	static AppDomain smethod_1()
	{
		return AppDomain.CurrentDomain;
	}

	static Assembly[] smethod_2(AppDomain appDomain_0)
	{
		return appDomain_0.GetAssemblies();
	}

	static AssemblyName smethod_3(Assembly assembly_0)
	{
		return assembly_0.GetName();
	}

	static string smethod_4(AssemblyName assemblyName_0)
	{
		return assemblyName_0.Name;
	}

	static bool smethod_5(string string_0, string string_1, StringComparison stringComparison_0)
	{
		return string.Equals(string_0, string_1, stringComparison_0);
	}

	static CultureInfo smethod_6(AssemblyName assemblyName_0)
	{
		return assemblyName_0.CultureInfo;
	}

	static void smethod_7(Stream stream_0, byte[] byte_0, int int_0, int int_1)
	{
		stream_0.Write(byte_0, int_0, int_1);
	}

	static int smethod_8(Stream stream_0, byte[] byte_0, int int_0, int int_1)
	{
		return stream_0.Read(byte_0, int_0, int_1);
	}

	static Assembly smethod_9()
	{
		return Assembly.GetExecutingAssembly();
	}

	static bool smethod_10(string string_0, string string_1)
	{
		return string_0.EndsWith(string_1);
	}

	static Stream smethod_11(Assembly assembly_0, string string_0)
	{
		return assembly_0.GetManifestResourceStream(string_0);
	}

	static DeflateStream smethod_12(Stream stream_0, CompressionMode compressionMode_0)
	{
		return new DeflateStream(stream_0, compressionMode_0);
	}

	static MemoryStream smethod_13()
	{
		return new MemoryStream();
	}

	static void smethod_14(Stream stream_0, long long_0)
	{
		stream_0.Position = long_0;
	}

	static void smethod_15(IDisposable idisposable_0)
	{
		idisposable_0.Dispose();
	}

	static long smethod_16(Stream stream_0)
	{
		return stream_0.Length;
	}

	static string smethod_17(string string_0)
	{
		return string_0.ToLowerInvariant();
	}

	static bool smethod_18(string string_0)
	{
		return string.IsNullOrEmpty(string_0);
	}

	static string smethod_19(string string_0, object object_0, object object_1)
	{
		return string.Format(string_0, object_0, object_1);
	}

	static Assembly smethod_20(byte[] byte_0, byte[] byte_1)
	{
		return Assembly.Load(byte_0, byte_1);
	}

	static Assembly smethod_21(byte[] byte_0)
	{
		return Assembly.Load(byte_0);
	}

	static AssemblyName smethod_22(string string_0)
	{
		return new AssemblyName(string_0);
	}

	static AssemblyNameFlags smethod_23(AssemblyName assemblyName_0)
	{
		return assemblyName_0.Flags;
	}

	static Assembly smethod_24(AssemblyName assemblyName_0)
	{
		return Assembly.Load(assemblyName_0);
	}

	static string smethod_25(ResolveEventArgs resolveEventArgs_0)
	{
		return resolveEventArgs_0.Name;
	}

	static void smethod_26(AppDomain appDomain_0, ResolveEventHandler resolveEventHandler_0)
	{
		appDomain_0.AssemblyResolve += resolveEventHandler_0;
	}
}
